presto-native-execution/presto_cpp/main/operators/PartitionAndSerialize.cpp

```diff
@@ -68,19 +73,135 @@ class PartitionAndSerializeOperator : public Operator {
+  // If replicateNullsAndAny_ is true, we need to duplicate ALL null values to
+  // ALL partition and also duplicate a non-null value to each empty partition.
+  // It is ok to over-allocate and then resize to a lower value instead of
+  // paying to reallocation.
+  vector_size_t calculateMaxOutputSize() {
+    const vector_size_t numInput = input_->size();
+    int32_t maxOutputSize = numInput;
+    if (replicateNullsAndAny_) {
+      const int32_t nullCount = nullRows_.countSelected();
+      // TODO: If numPartitions_ and nullCount is large, we need to batch out
+      // getOutput logic
+      maxOutputSize = numInput + (nullCount * numPartitions_) + numPartitions_;
+    }
+    return maxOutputSize;
+  }
+
+  // If replicateNullsAndAny_ is true, we need to duplicate null values across
+  // all partition ids. For efficiency, we calculate how many nulls needs to be
+  // added to each partition id.
+  // Returns the total elements in partitionVector
+  vector_size_t fillDataWithNull(
+      FlatVector<StringView>& dataVector,
+      FlatVector<int32_t> partitionVector) {
+    auto numInput = input_->size();
+    auto insertInto = numInput;
+
+    for (auto forPartition = 0; forPartition < numPartitions_; ++forPartition) {
+      for (auto copyFromIdx = 0; copyFromIdx < numInput; ++copyFromIdx) {
+        // Copy only if copyFromIdx has null partition key and it is not already
+        // assigned to forPartition.
+        // If Partition key is null, copy the serialised data into target index
+        if (nullRows_.isValid(copyFromIdx) &&
+            partitions_[copyFromIdx] != forPartition) {
+          dataVector.setNoCopy(insertInto, dataVector.valueAt(copyFromIdx));
+          partitionVector.set(insertInto, forPartition);
+          insertInto++;
+        }
+      }
+    }
+    return insertInto;
+  }
+
+  // Identify the first non-empty element and replicate on all empty partitions
+  vector_size_t replicateAnyInEmptyPartitions(
+      FlatVector<int32_t>& partitionsVector,
+      FlatVector<StringView>& dataVector,
+      const vector_size_t start) {
+    auto rawPartitions = partitionsVector.mutableRawValues();
+    const auto numInput = input_->size();
+    std::set<int32_t> nonEmptyPartitions(
+        rawPartitions, rawPartitions + numInput);
+    StringView firstNonEmpty;
+    for (int idx = 0; idx < numInput; ++idx) {
+      if (!nullRows_.isValid(idx)) {
+        firstNonEmpty = dataVector.valueAt(idx);
+        break;
+      }
+    }
+
+    int32_t finalOutputSize = start;
+    for (auto idx = 0; idx < numPartitions_; ++idx) {
+      if (nonEmptyPartitions.find(idx) == nonEmptyPartitions.end()) {
+        rawPartitions[finalOutputSize] = idx;
+        dataVector.setNoCopy(finalOutputSize, firstNonEmpty);
+        finalOutputSize++;
+      }
+    }
+    return finalOutputSize;
+  }
+
+  // A partition key is considered null only if its recursive children all are
```

Comment:
```
// A partition key is considered null if any of the partition key column in 'input_' is null. The function records the rows has null partition key in 'nullRows_'.
```

Reply:
Rephrased the comment, but with a correction
> considered null only if its recursive children all are null
instead of
> any of the partition key column in 'input_' is null

Reply:
Yes, all the children needs to be null. This is handled in `.decode()` method.

Reply:
orWithNegatedBits set null bit of a row if any of its children is null